jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java

```diff
@@ -673,11 +684,109 @@ public void write(int b) throws IOException
 
     @Override
     public void print(String s) throws IOException
+    {
+        print(s,false);
+    }
+
+    @Override
+    public void println(String s) throws IOException
+    {
+        print(s,true);
+    }
+
+    private void print(String s, boolean eoln) throws IOException
     {
         if (isClosed())
             throw new IOException("Closed");
 
-        write(s.getBytes(_channel.getResponse().getCharacterEncoding()));
+        // TODO would a Threadlocal pool be better?
+        String charset = _channel.getResponse().getCharacterEncoding();
+        if (_encoder==null || !_encoder.charset().name().equals(charset))
+            _encoder = Charset.forName(charset).newEncoder();
+        else
+            _encoder.reset();
+
+        CharBuffer in = CharBuffer.wrap(s);
+        CharBuffer crlf = eoln?CharBuffer.wrap("\r\n"):null;
+        ByteBuffer out = getHttpChannel().getByteBufferPool().acquire((int)(1+(s.length()+2)*_encoder.averageBytesPerChar()),false);
+        BufferUtil.flipToFill(out);
+
+        for(;;)
+        {
+            CoderResult result;
+            if (in.hasRemaining())
+            {
+                result = _encoder.encode(in, out, false);
+                if (result.isUnderflow())
+                    if (crlf==null)
+                        break;
+                    else
+                        continue;
+            }
+            else if (crlf.hasRemaining())
+            {
+                result = _encoder.encode(crlf, out, true);
+                if (result.isUnderflow())
+                {
+                    if (!_encoder.flush(out).isUnderflow())
+                        result.throwException();
+                    break;
+                }
+            }
+            else
+                break;
+
+            if (result.isOverflow())
+            {
+                BufferUtil.flipToFlush(out,0);
+                ByteBuffer bigger = BufferUtil.ensureCapacity(out,out.capacity()+s.length()+2);
+                getHttpChannel().getByteBufferPool().release(out);
+                BufferUtil.flipToFill(bigger);
+                out = bigger;
+                continue;
+            }
+
+            result.throwException();
+        }
+        BufferUtil.flipToFlush(out,0);
+        write(out.array(),out.arrayOffset(),out.remaining());
```

Comment:
We have a `ByteBuffer`, why using the `write()` for arrays? Just write the `ByteBuffer`, no?

Reply:
Because for whatever reason, the write(ByteBuffer) implementation bypasses the aggregation buffer, so to use that method is a change of behaviour and a break of the servlet spec.